src/java.base/windows/classes/sun/nio/fs/WindowsFileSystemProvider.java

```diff
@@ -418,65 +419,137 @@ public void checkAccess(Path obj, AccessMode... modes) throws IOException {
         }
     }
 
-    @Override
-    public boolean isSameFile(Path obj1, Path obj2) throws IOException {
-        WindowsPath file1 = WindowsPath.toWindowsPath(obj1);
-        if (file1.equals(obj2))
-            return true;
-        if (obj2 == null)
-            throw new NullPointerException();
-        if (!(obj2 instanceof WindowsPath))
+    // contains file attributes and its handle
+    // the attributes' key is valid as long as the handle remains open
+    private record LinkAttributes(WindowsFileAttributes attrs, long handle) {
+        public boolean equals(Object obj) {
+            if (obj == this)
+                return true;
+            if (obj instanceof LinkAttributes other)
+                return WindowsFileAttributes.isSameFile(attrs, other.attrs());
             return false;
-        WindowsPath file2 = (WindowsPath)obj2;
+        }
 
-        // open both files and see if they are the same
-        long h1 = 0L;
+        public int hashCode() {
+            return attrs.volSerialNumber() +
+                   attrs.fileIndexHigh() + attrs.fileIndexLow();
+        }
+    }
+
+    // find the attributes of the last accessible link in the chain
+    private LinkAttributes lastFileAttrs(WindowsPath path)
+        throws IOException, WindowsException
+    {
+        var fileAttrs = new HashSet<LinkAttributes>();
+        LinkAttributes lastFileAttrs = null;
         try {
-            h1 = file1.openForReadAttributeAccess(true);
-        } catch (WindowsException x) {
-            if (x.lastError() != ERROR_FILE_NOT_FOUND &&
-                x.lastError() != ERROR_PATH_NOT_FOUND)
-                x.rethrowAsIOException(file1);
+            while (path != null) {
+                long h;
+                try {
+                    h = path.openForReadAttributeAccess(false);
+                } catch (WindowsException x) {
+                    if (x.lastError() != ERROR_FILE_NOT_FOUND &&
+                        x.lastError() != ERROR_PATH_NOT_FOUND)
+                        throw x;
+                    break;
+                }
+
+                WindowsFileAttributes attrs =
+                    WindowsFileAttributes.readAttributes(h);
+                if (!attrs.isSymbolicLink())
+                    break;
+
+                LinkAttributes linkAttr = new LinkAttributes(attrs, h);
+                if (!fileAttrs.add(linkAttr))
+                    throw new FileSystemLoopException("Looping symbolic link");
```

Comment:
The parameter to FileSystemLoopException is the file that caused the loop rather than a message.

Reply:
Fixed in bd4b94d.